content/posts/optree/pytrees/index.md

```diff
@@ -0,0 +1,203 @@
+---
+title: "Pytrees for Scientific Python"
+date: 2025-05-14T10:27:59-07:00
+draft: false
+description: "
+Introducing PyTrees for Scientific Python. We discuss what PyTrees are, how they're useful in the realm of scientific python, and how to work _efficiently_ with them.
+"
+tags: ["PyTrees", "Functional Programming", "Tree-like data manipulation"]
+displayInList: true
+author: ["Peter Fackeldey", "Mihai Maruseac", "Matthew Feickert"]
+---
+
+## Manipulating Tree-like Data using Functional Programming Paradigms
+
+A "PyTree" is a nested collection of Python containers (e.g. dicts, (named) tuples, lists, ...), where the leafs are of interest.
+As you can imagine (or even experienced in the past), such arbitrary nested collections can be cumbersome to manipulate _efficiently_.
+It often requires complex recursive logic which usually does not generalize to other nested Python containers (PyTrees).
+
+The core concept of PyTrees is being able to flatten them into a flat collection of leafs and a "blueprint" of the tree structure, and then being able to unflatten them back into the original PyTree.
+This allows to apply generic transformations, e.g. taking the square root of each leaf of a PyTree with a `tree_map(np.sqrt, pytree)` operation:
+
+```python
+import optree as pt
+import numpy as np
+
+# tuple of a list of a dict with an array as value, and an array
+pytree = ([[{"foo": np.array([4.0])}], np.array([9.0])],)
+
+# sqrt of each leaf array
+sqrt_pytree = pt.tree_map(np.sqrt, pytree)
+print(f"{sqrt_pytree=}")
+# >> sqrt_pytree=([[{'foo': array([2.])}], array([3.])],)
+
+# reductions
+all_positive = pt.tree_all(pt.tree_map(lambda x: x > 0.0, pytree))
+print(f"{all_positive=}")
+# >> all_positive=True
+
+summed = pt.tree_reduce(sum, pytree)
+print(f"{summed=}")
+# >> summed=array([13.])
+```
+
+The trick here is that these operations can be implemented in three steps, e.g. `tree_map`:
+
+```python
+# step 1:
+leafs, treedef = pt.tree_flatten(pytree)
+
+# step 2:
+new_leafs = tuple(map(fun, leafs))
+
+# step 3:
+result_pytree = pt.tree_unflatten(treedef, new_leafs)
+```
+
+Here, we use [`optree`](https://github.com/metaopt/optree/tree/main/optree) - a standalone PyTree library - that enables all these manipulations. It focusses on performance, feature richness, minimal dependencies, and got adopted by [PyTorch](https://pytorch.org), [Keras](https://keras.io), and [TensorFlow](https://github.com/tensorflow/tensorflow) (through Keras) as a core dependency.
```

Comment:
```suggestion
Here, we use [`optree`](https://github.com/metaopt/optree/tree/main/optree) &mdash; a standalone PyTree library &mdash; that enables all these manipulations. It focuses on performance, is feature rich, has minimal dependencies, and has been adopted by [PyTorch](https://pytorch.org), [Keras](https://keras.io), and [TensorFlow](https://github.com/tensorflow/tensorflow) (through Keras) as a core dependency.
```

* Use em dashes (&mdash;) instead of hyphens (-)
* Small typo fix: `focusses` -> `focuses`
* Possible rewording of sentence.